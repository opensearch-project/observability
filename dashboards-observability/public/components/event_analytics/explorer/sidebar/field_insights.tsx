/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useMemo, useState, useContext, useEffect } from 'react';
import { indexOf, last } from 'lodash';
import { EuiFlexGroup, EuiFlexItem, EuiLink, EuiBasicTable } from '@elastic/eui';
import { getIndexPatternFromRawQuery } from '../../../../../common/utils/query_utils';
import { TabContext } from '../../hooks/use_tab_context';

export const FieldInsights = ({ field, query }: any) => {
  const { pplService } = useContext(TabContext);
  const { rawQuery } = query;
  const index = getIndexPatternFromRawQuery(rawQuery);
  const generalReports = [
    {
      id: 'top_values',
      name: 'Top values',
      query: `source = ${index} | top 5 ${field.name} | sort - ${field.name}`,
      format: 'jdbc',
    },
    {
      id: 'rare_values',
      name: 'Rare values',
      query: `source = ${index} | rare ${field.name} | sort + ${field.name}`,
      format: 'jdbc',
    },
  ];

  const numericalOnlyReports = [
    {
      id: 'average',
      name: 'Average overtime',
      query: `source = ${index} | stats avg(${field.name})`,
      format: 'viz',
    },
    {
      id: 'maximum',
      name: 'Maximum overtime',
      query: `source = ${index} | stats max(${field.name})`,
      format: 'viz',
    },
    {
      id: 'minimum',
      name: 'Minimum overtime',
      query: `source = ${index} | stats min(${field.name})`,
      format: 'viz',
    },
  ];
  const NUMERICAL_TYPES = ['short', 'integer', 'long', 'float', 'double'];
  const isNumericalField = indexOf(NUMERICAL_TYPES, field.type) > 0;
  const [curReport, setCurReport] = useState({ ...generalReports[0] });
  const [reportContent, setReportContent] = useState({});

  const statsInsightsQueries = [
    {
      id: 'stats',
      name: 'Stats',
      query: `source = ${index} | stats avg(${field.name}), max(${field.name}), min(${field.name})`,
      format: 'viz',
    },
  ];

  const fetchData = async (requests) => {
    return await Promise.all(
      requests.map((reqQuery) => {
        const req = {
          format: reqQuery.format,
          query: reqQuery.query,
        };
        return getInsights(req);
      })
    );
  };

  useEffect(() => {
    let requests = [...generalReports];
    if (isNumericalField) requests = [...requests, ...statsInsightsQueries];
    fetchData(requests)
      .then((res) => {
        // numerical field
        generalReports.map((report, index) => {
          if (!res[index]?.jsonData) return;
          setReportContent((staleState) => {
            return {
              ...staleState,
              [report.id]: res[index]?.jsonData || {},
            };
          });
        });
        if (res.length > 2) {
          const statsRes = last(res);
          if (!statsRes?.metadata) return;
          numericalOnlyReports.map((rep, index) => {
            const fieldName = statsRes.metadata?.fields[index]?.name;
            setReportContent((staleState) => {
              return {
                ...staleState,
                [rep.id]: [{ [field.name]: statsRes.data[fieldName][0] }],
              };
            });
          });
        }
      })
      .catch((error) => console.error(error));
  }, []);

  const getInsights = async (query: string) => {
    try {
      return await pplService.fetch(query);
    } catch (error) {
      console.error(error);
    }
  };

  const insightsContent = useMemo(() => {
    const columns = [
      {
        field: field.name,
        name: field.name,
      },
    ];
    const repItems = reportContent[curReport.id] || [];

    return <EuiBasicTable columns={columns} items={repItems} />;
  }, [curReport, reportContent, field.name]);

  return (
    <EuiFlexGroup direction="column" data-test-subj="sidebarField__fieldInsights">
      <EuiFlexItem grow={false}>
        <EuiFlexGroup wrap>
          {generalReports.map((report) => {
            return (
              <EuiFlexItem grow={false}>
                <EuiLink onClick={() => setCurReport(report)}>{report.name}</EuiLink>
              </EuiFlexItem>
            );
          })}
          {indexOf(NUMERICAL_TYPES, field.type) > 0 &&
            numericalOnlyReports.map((report) => {
              return (
                <EuiFlexItem grow={false}>
                  <EuiLink onClick={() => setCurReport(report)}>{report.name}</EuiLink>
                </EuiFlexItem>
              );
            })}
        </EuiFlexGroup>
      </EuiFlexItem>
      <EuiFlexItem grow={false}>{insightsContent}</EuiFlexItem>
    </EuiFlexGroup>
  );
};
