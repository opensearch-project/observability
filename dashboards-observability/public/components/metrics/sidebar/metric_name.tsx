/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { EuiButtonEmpty, EuiToken } from '@elastic/eui';

interface IMetricNameProps {
  metric: any;
  handleClick: (props: any) => void;
}

export const MetricName = (props: IMetricNameProps) => {
  const { metric, handleClick } = props;

  const title = metric.catalog === 'CUSTOM_METRICS' ? 'OpenSearch' : 'Prometheus';
  const token =
    metric.catalog === 'CUSTOM_METRICS'
      ? '/ui/default_branding/opensearch_mark_default_mode.svg'
      : 'tokenProperty';
  const name = (metricName: string) => {
    return metric.catalog === 'CUSTOM_METRICS' ? metricName : metricName.split('.')[1];
  };

  return (
    <EuiButtonEmpty
      className="eui-textTruncate"
      title={metric.name}
      onClick={() => handleClick(metric)}
    >
      <EuiToken
        className="tokenMargin"
        title={title}
        iconType={token}
        fill="light"
        color="euiColorVis1"
        shape="square"
      />{' '}
      {name(metric.name)}
    </EuiButtonEmpty>
  );
};
