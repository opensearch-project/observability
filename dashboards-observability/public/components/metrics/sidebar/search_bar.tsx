/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiGlobalToastList, EuiSearchBar, EuiToast } from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { pplServiceRequestor } from '../helpers/utils';

interface ISearchBarProps {
  allAvailableMetrics: any;
  handleAddMetric: any;
}

export const SearchBar = (props: ISearchBarProps) => {
  const { allAvailableMetrics, handleAddMetric } = props;

  const [query, setQuery] = useState('');
  const [error, setError] = useState(null);

  const [toasts, setToasts] = useState([]);
  const addToast = (res: string) => {
    if (res === 'success') {
      const toast = {
        id: 'success',
        title: 'Metric successfully added!!',
        color: 'success',
      };
      setToasts(toasts.concat(toast));
    } else {
      const toast = {
        id: 'fail',
        title: 'Metric not found.',
        color: 'danger',
      };
      setToasts(toasts.concat(toast));
    }
  };
  const removeToast = (removedToast: any) => {
    setToasts(toasts.filter((toast: any) => toast.id !== removedToast.id));
  };

  const onChange = ({ query }) => {
    const metric = allAvailableMetrics.find((row: any) => row.name.includes(query.text));
    if (metric) {
      handleAddMetric(metric);
      addToast('success');
    } else {
      addToast('fail');
    }
  };

  return (
    <div>
      <EuiSearchBar defaultQuery={''} onChange={onChange} />
      <EuiGlobalToastList toasts={toasts} dismissToast={removeToast} toastLifeTimeMs={6000} />
    </div>
  );
};
