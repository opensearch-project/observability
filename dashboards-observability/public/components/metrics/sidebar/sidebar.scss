/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

.dscSidebar__container {
  padding-left: 0 !important;
  padding-right: 0 !important;
  background-color: transparent;
  border-right-color: transparent;
  border-bottom-color: transparent;
}

.dscIndexPattern__container {
  display: flex;
  align-items: center;
  height: $euiSize * 3;
  margin-top: -$euiSizeS;
}

.dscIndexPattern__triggerButton {
  @include euiTitle('xs');
  line-height: $euiSizeXXL;
}

.dscFieldList {
  list-style: none;
  margin-bottom: 0;
}

.dscFieldListHeader {
  padding: $euiSizeS $euiSizeS 0 $euiSizeS;
  background-color: lightOrDarkTheme(tint($euiColorPrimary, 90%), $euiColorLightShade);
}

.dscFieldList--popular {
  background-color: lightOrDarkTheme(tint($euiColorPrimary, 90%), $euiColorLightShade);
}

.dscFieldChooser__toggle {
  color: $euiColorMediumShade;
  margin-left: $euiSizeS  !important;
}

/**
 * 1. Only visually hide the action, so that it's still accessible to screen readers.
 * 2. When tabbed to, this element needs to be visible for keyboard accessibility.
 */
.dscSidebarItem__action {
  opacity: 0;
  /* 1 */
  transition: none;

  &:focus {
    opacity: 1;
    /* 2 */
  }

  font-size: $euiFontSizeXS;
  padding: 2px 6px !important;
  height: 22px !important;
  min-width: auto !important;

  .euiButton__content {
    padding: 0 4px;
  }
}

.dscFieldSearch {
  padding: $euiSizeS;
}

.dscFieldSearch__toggleButton {
  width: calc(100% - #{$euiSizeS});
  color: $euiColorPrimary;
  padding-left: $euiSizeXS;
  margin-left: $euiSizeXS;
}

.dscFieldSearch__filterWrapper {
  flex-grow: 0;
}

.dscFieldSearch__formWrapper {
  padding: $euiSizeM;
}

.dscFieldDetails {
  color: $euiTextColor;
  margin-bottom: $euiSizeS;
}

.dscSidebarItem__fieldPopoverPanel {
  min-width: 300px;
  max-width: 600px;
  max-height: 600px;
  overflow-y: scroll;
}

.override_timestamp_loading {
  vertical-align: middle;
}

.explorerFieldSelector {
  padding: $euiSizeS;
}

#vis__mainContent {
  .explorer__insights {
    min-height: 0;
    display: grid;
    grid-template-columns: 50% 50%;
    height: 100%;

    .explorerFieldSelector,
    .explorer__vizDataConfig {
      padding: $euiSizeS;
      overflow: auto;

      &__fieldGroups {
        @include euiYScrollWithShadows;

        overflow-y: auto;
        margin-right: -$euiSizeS;
        padding-right: $euiSizeS;
        margin-top: $euiSizeS;
      }

      &__fieldGroup {
        margin-top: $euiSizeS;

        &:first-child {
          margin-top: 0;
        }
      }
    }

    .explorerFieldSelector {
      @include euiYScrollWithShadows;

      .sidebar-list {
        height: 100%;
      }
    }
  }
}

.sidebarHeight {
  line-height: normal;
  overflow: auto;
  max-height: 100vh;
}

.metricsList {
  line-height: normal;
}

.metricsListContainer {
  margin-left: 15px;
  margin-right: 5px;
}

.tokenMargin {
  vertical-align: middle;
}