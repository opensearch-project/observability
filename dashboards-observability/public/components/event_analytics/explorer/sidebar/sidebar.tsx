/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import './sidebar.scss';

import React, { useState } from 'react';
import { isEmpty } from 'lodash';
import { EuiTitle, EuiSpacer, EuiFieldSearch, EuiAccordion } from '@elastic/eui';
import { I18nProvider } from '@osd/i18n/react';
import { Field } from './field';
import { IExplorerFields, IField } from '../../../../../common/types/explorer';

interface ISidebarProps {
  query: string;
  explorerFields: IExplorerFields;
  explorerData: any;
  selectedPattern: string;
  isOverridingPattern: boolean;
  selectedTimestamp: string;
  isOverridingTimestamp: boolean;
  isFieldToggleButtonDisabled: boolean;
  handleOverridePattern: (pattern: IField) => void;
  handleOverrideTimestamp: (timestamp: IField) => void;
  handleAddField: (field: IField) => void;
  handleRemoveField: (field: IField) => void;
}

export const Sidebar = (props: ISidebarProps) => {
  const {
    query,
    explorerFields,
    explorerData,
    selectedPattern,
    isOverridingPattern,
    selectedTimestamp,
    isOverridingTimestamp,
    isFieldToggleButtonDisabled,
    handleOverridePattern,
    handleOverrideTimestamp,
    handleAddField,
    handleRemoveField,
  } = props;

  const [showFields, setShowFields] = useState<boolean>(false);
  const [searchTerm, setSearchTerm] = useState<string>('');

  return (
    <I18nProvider>
      <section className="sidebar-list">
        <div className="dscSidebar__item">
          <EuiFieldSearch
            compressed
            fullWidth
            onChange={(e) => {
              setSearchTerm(e.target.value);
            }}
            placeholder="Search field names"
            value={searchTerm}
            data-test-subj="eventExplorer__sidebarSearch"
          />
        </div>
        <EuiSpacer size="s" />
        <div className="sidebar-list">
          {((explorerData && !isEmpty(explorerData.jsonData) && !isEmpty(explorerFields)) ||
            !isEmpty(explorerFields.availableFields)) && (
            <>
              {explorerFields?.queriedFields && explorerFields.queriedFields?.length > 0 && (
                <EuiAccordion
                  initialIsOpen
                  id="fieldSelector__queriedFields"
                  buttonContent={
                    <EuiTitle size="xxxs">
                      <span>Query fields</span>
                    </EuiTitle>
                  }
                  paddingSize="xs"
                >
                  <ul
                    className="dscSidebarList dscFieldList--selected"
                    aria-labelledby="queried_fields"
                    data-test-subj={`fieldList-selected`}
                  >
                    {explorerFields.queriedFields &&
                      explorerFields.queriedFields.map((field) => {
                        return (
                          <li
                            key={`field${field.name}`}
                            data-attr-field={field.name}
                            className="dscSidebar__item"
                          >
                            <Field
                              query={query}
                              field={field}
                              selectedPattern={selectedPattern}
                              isOverridingPattern={isOverridingPattern}
                              handleOverridePattern={handleOverridePattern}
                              isOverridingTimestamp={isOverridingTimestamp}
                              selectedTimestamp={selectedTimestamp}
                              handleOverrideTimestamp={handleOverrideTimestamp}
                              selected={true}
                              isFieldToggleButtonDisabled={true}
                              showTimestampOverrideButton={false}
                              onToggleField={handleRemoveField}
                            />
                          </li>
                        );
                      })}
                  </ul>
                </EuiAccordion>
              )}
              <EuiSpacer size="s" />
              <EuiAccordion
                initialIsOpen
                id="fieldSelector__selectedFields"
                buttonContent={
                  <EuiTitle size="xxxs">
                    <span>Selected Fields</span>
                  </EuiTitle>
                }
                paddingSize="xs"
              >
                <ul
                  className="dscSidebarList dscFieldList--selected"
                  aria-labelledby="selected_fields"
                  data-test-subj={`fieldList-selected`}
                >
                  {explorerData &&
                    !isEmpty(explorerData.jsonData) &&
                    explorerFields.selectedFields &&
                    explorerFields.selectedFields.map((field) => {
                      return (
                        <li
                          key={`field${field.name}`}
                          data-attr-field={field.name}
                          className="dscSidebar__item"
                        >
                          <Field
                            query={query}
                            field={field}
                            selectedPattern={selectedPattern}
                            isOverridingPattern={isOverridingPattern}
                            handleOverridePattern={handleOverridePattern}
                            selectedTimestamp={selectedTimestamp}
                            isOverridingTimestamp={isOverridingTimestamp}
                            handleOverrideTimestamp={handleOverrideTimestamp}
                            selected={true}
                            isFieldToggleButtonDisabled={isFieldToggleButtonDisabled}
                            showTimestampOverrideButton={true}
                            onToggleField={handleRemoveField}
                          />
                        </li>
                      );
                    })}
                </ul>
              </EuiAccordion>
              <EuiSpacer size="s" />
              <EuiAccordion
                initialIsOpen
                id="fieldSelector__availableFields"
                buttonContent={
                  <EuiTitle size="xxxs">
                    <span>Available Fields</span>
                  </EuiTitle>
                }
                paddingSize="xs"
              >
                <ul
                  className={`dscFieldList dscFieldList--unpopular ${
                    !showFields ? 'hidden-sm hidden-xs' : ''
                  }`}
                  aria-labelledby="available_fields"
                  data-test-subj={`fieldList-unpopular`}
                >
                  {explorerFields.availableFields &&
                    explorerFields.availableFields
                      .filter((field) => searchTerm === '' || field.name.indexOf(searchTerm) !== -1)
                      .map((field) => {
                        return (
                          <li
                            key={`field${field.name}`}
                            data-attr-field={field.name}
                            className="dscSidebar__item"
                          >
                            <Field
                              query={query}
                              field={field}
                              selectedPattern={selectedPattern}
                              isOverridingPattern={isOverridingPattern}
                              handleOverridePattern={handleOverridePattern}
                              selectedTimestamp={selectedTimestamp}
                              isOverridingTimestamp={isOverridingTimestamp}
                              handleOverrideTimestamp={handleOverrideTimestamp}
                              onToggleField={handleAddField}
                              selected={false}
                              isFieldToggleButtonDisabled={isFieldToggleButtonDisabled}
                              showTimestampOverrideButton={true}
                            />
                          </li>
                        );
                      })}
                </ul>
              </EuiAccordion>
            </>
          )}
        </div>
      </section>
    </I18nProvider>
  );
};
