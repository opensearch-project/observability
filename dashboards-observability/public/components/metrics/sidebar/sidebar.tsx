/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import './sidebar.scss';

import React, { useEffect } from 'react';
import { EuiTitle, EuiSpacer, EuiAccordion, EuiLink } from '@elastic/eui';
import { I18nProvider } from '@osd/i18n/react';
import { batch, useDispatch, useSelector } from 'react-redux';
import {
  availableMetricsSelector,
  deSelectMetric,
  selectMetric,
  loadMetrics,
  selectedMetricsSelector,
  recentlyCreatedMetricsSelector,
} from '../redux/slices/metrics_slice';
import { CoreStart } from '../../../../../../src/core/public';
import PPLService from '../../../services/requests/ppl';

interface ISidebarProps {
  http: CoreStart['http'];
  pplService: PPLService;
}

export const Sidebar = (props: ISidebarProps) => {
  const { http, pplService } = props;
  const dispatch = useDispatch();

  const availableMetrics = useSelector(availableMetricsSelector);
  const selectedMetrics = useSelector(selectedMetricsSelector);
  const recentlyCreatedMetrics = useSelector(recentlyCreatedMetricsSelector);

  useEffect(() => {
    batch(() => {
      dispatch(loadMetrics({ http, pplService }));
    });
  }, []);

  const handleAddMetric = (metric: any) => dispatch(selectMetric(metric));

  const handleRemoveMetric = (metric: any) => {
    dispatch(deSelectMetric(metric));
  };

  return (
    <I18nProvider>
      <section className="sidebarHeight">
        <EuiAccordion
          initialIsOpen
          id="recentlyCreatedMetricsSelector"
          buttonContent={
            <EuiTitle size="xxxs">
              <span>Recently Created Metrics</span>
            </EuiTitle>
          }
          paddingSize="none"
        >
          <ul className="metricsList">
            {recentlyCreatedMetrics.map((metric: any) => (
              <li key={metric.id} className="metricsListContainer">
                <EuiLink
                  className="metricName eui-textTruncate"
                  title={metric.name}
                  onClick={() => handleAddMetric(metric)}
                >
                  {metric.name}
                </EuiLink>
              </li>
            ))}
          </ul>
        </EuiAccordion>
        <EuiSpacer size="s" />
        <EuiAccordion
          initialIsOpen
          id="selectedMetricsSelector"
          buttonContent={
            <EuiTitle size="xxxs">
              <span>Selected Metrics</span>
            </EuiTitle>
          }
          paddingSize="none"
        >
          <ul className="metricsList">
            {selectedMetrics.map((metric: any) => (
              <li key={metric.id} className="metricsListContainer">
                <EuiLink
                  className="metricName eui-textTruncate"
                  title={metric.name}
                  onClick={() => handleRemoveMetric(metric)}
                >
                  {metric.name}
                </EuiLink>
              </li>
            ))}
          </ul>
        </EuiAccordion>
        <EuiSpacer size="s" />
        <EuiAccordion
          initialIsOpen
          id="availableMetricsSelector"
          buttonContent={
            <EuiTitle size="xxxs">
              <span>Available Metrics</span>
            </EuiTitle>
          }
          paddingSize="none"
        >
          <ul className="metricsList">
            {availableMetrics.slice(0, 100).map((metric: any) => (
              <li key={metric.id} className="metricsListContainer">
                <EuiLink
                  className="metricName eui-textTruncate"
                  title={metric.name}
                  onClick={() => handleAddMetric(metric)}
                >
                  {metric.name}
                </EuiLink>
              </li>
            ))}
          </ul>
          {availableMetrics.length > 100 && (
            <p>Use search bar for searching through all metrics.</p>
          )}
        </EuiAccordion>
      </section>
    </I18nProvider>
  );
};
