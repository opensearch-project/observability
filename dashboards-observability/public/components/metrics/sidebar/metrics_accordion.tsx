/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { EuiAccordion, EuiTitle } from '@elastic/eui';
import { MetricName } from './metric_name';

interface IMetricNameProps {
  metricsList: [];
  headerName: string;
  handleClick: (props: any) => void;
}

export const MetricsAccordion = (props: IMetricNameProps) => {
  const { metricsList, headerName, handleClick } = props;

  return (
    <EuiAccordion
      initialIsOpen
      id={`${headerName}Selector`}
      buttonContent={
        <EuiTitle size="xxxs">
          <span>{headerName}</span>
        </EuiTitle>
      }
      paddingSize="none"
    >
      <ul className="metricsList">
        {metricsList.slice(0, 100).map((metric: any) => (
          <li key={metric.id} className="metricsListContainer">
            <MetricName metric={metric} handleClick={handleClick} />
          </li>
        ))}
      </ul>
      {metricsList.length > 100 && <p>Use search bar for searching through all metrics.</p>}
    </EuiAccordion>
  );
};
