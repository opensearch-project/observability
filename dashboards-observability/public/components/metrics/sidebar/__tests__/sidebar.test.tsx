/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import httpClientMock from '../../../../../test/__mocks__/httpClientMock';
import PPLService from '../../../../services/requests/ppl';
import { applyMiddleware, createStore } from '@reduxjs/toolkit';
import rootReducer from '../../../../framework/redux/reducers';
import { Provider } from 'react-redux';
import { Sidebar } from '../sidebar';
import thunk from 'redux-thunk';

describe('Side Bar Component', () => {
  configure({ adapter: new Adapter() });
  const store = createStore(rootReducer, applyMiddleware(thunk));

  it('renders Side Bar Component', async () => {
    httpClientMock.get = jest.fn();

    const http = httpClientMock;
    const pplService = new PPLService(httpClientMock);
    const search = false;

    const wrapper = mount(
      <Provider store={store}>
        <Sidebar http={http} pplService={pplService} search={search} />
      </Provider>
    );

    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
