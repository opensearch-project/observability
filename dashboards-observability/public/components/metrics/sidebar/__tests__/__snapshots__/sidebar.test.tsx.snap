// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Side Bar Component renders Side Bar Component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    }
  }
>
  <Sidebar
    http={[MockFunction]}
    pplService={
      PPLService {
        "fetch": [Function],
        "http": [MockFunction],
      }
    }
  >
    <I18nProvider>
      <IntlProvider
        defaultLocale="en"
        formats={
          Object {
            "date": Object {
              "full": Object {
                "day": "numeric",
                "month": "long",
                "weekday": "long",
                "year": "numeric",
              },
              "long": Object {
                "day": "numeric",
                "month": "long",
                "year": "numeric",
              },
              "medium": Object {
                "day": "numeric",
                "month": "short",
                "year": "numeric",
              },
              "short": Object {
                "day": "numeric",
                "month": "numeric",
                "year": "2-digit",
              },
            },
            "number": Object {
              "currency": Object {
                "style": "currency",
              },
              "percent": Object {
                "style": "percent",
              },
            },
            "relative": Object {
              "days": Object {
                "units": "day",
              },
              "hours": Object {
                "units": "hour",
              },
              "minutes": Object {
                "units": "minute",
              },
              "months": Object {
                "units": "month",
              },
              "seconds": Object {
                "units": "second",
              },
              "years": Object {
                "units": "year",
              },
            },
            "time": Object {
              "full": Object {
                "hour": "numeric",
                "minute": "numeric",
                "second": "numeric",
                "timeZoneName": "short",
              },
              "long": Object {
                "hour": "numeric",
                "minute": "numeric",
                "second": "numeric",
                "timeZoneName": "short",
              },
              "medium": Object {
                "hour": "numeric",
                "minute": "numeric",
                "second": "numeric",
              },
              "short": Object {
                "hour": "numeric",
                "minute": "numeric",
              },
            },
          }
        }
        locale="en"
        messages={Object {}}
        textComponent={Symbol(react.fragment)}
      >
        <PseudoLocaleWrapper>
          <section
            className="sidebarHeight"
          >
            <EuiAccordion
              arrowDisplay="left"
              buttonContent={
                <EuiTitle
                  size="xxxs"
                >
                  <span>
                    Recently Created Metrics
                  </span>
                </EuiTitle>
              }
              id="recentlyCreatedMetricsSelector"
              initialIsOpen={true}
              isLoading={false}
              isLoadingMessage={false}
              paddingSize="none"
            >
              <div
                className="euiAccordion euiAccordion-isOpen"
              >
                <div
                  className="euiAccordion__triggerWrapper"
                >
                  <button
                    aria-controls="recentlyCreatedMetricsSelector"
                    aria-expanded={true}
                    className="euiAccordion__button"
                    id="random_html_id"
                    onClick={[Function]}
                    type="button"
                  >
                    <span
                      className="euiAccordion__iconWrapper"
                    >
                      <EuiIcon
                        className="euiAccordion__icon euiAccordion__icon-isOpen"
                        size="m"
                        type="arrowRight"
                      >
                        <EuiIconEmpty
                          aria-hidden={true}
                          className="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
                          focusable="false"
                          role="img"
                          style={null}
                        >
                          <svg
                            aria-hidden={true}
                            className="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
                            focusable="false"
                            height={16}
                            role="img"
                            style={null}
                            viewBox="0 0 16 16"
                            width={16}
                            xmlns="http://www.w3.org/2000/svg"
                          />
                        </EuiIconEmpty>
                      </EuiIcon>
                    </span>
                    <span
                      className="euiIEFlexWrapFix"
                    >
                      <EuiTitle
                        size="xxxs"
                      >
                        <span
                          className="euiTitle euiTitle--xxxsmall"
                        >
                          Recently Created Metrics
                        </span>
                      </EuiTitle>
                    </span>
                  </button>
                </div>
                <div
                  aria-labelledby="random_html_id"
                  className="euiAccordion__childWrapper"
                  id="recentlyCreatedMetricsSelector"
                  role="region"
                  tabIndex={-1}
                >
                  <EuiResizeObserver
                    onResize={[Function]}
                  >
                    <div>
                      <div
                        className=""
                      >
                        <ul
                          className="metricsList"
                        />
                      </div>
                    </div>
                  </EuiResizeObserver>
                </div>
              </div>
            </EuiAccordion>
            <EuiSpacer
              size="s"
            >
              <div
                className="euiSpacer euiSpacer--s"
              />
            </EuiSpacer>
            <EuiAccordion
              arrowDisplay="left"
              buttonContent={
                <EuiTitle
                  size="xxxs"
                >
                  <span>
                    Selected Metrics
                  </span>
                </EuiTitle>
              }
              id="selectedMetricsSelector"
              initialIsOpen={true}
              isLoading={false}
              isLoadingMessage={false}
              paddingSize="none"
            >
              <div
                className="euiAccordion euiAccordion-isOpen"
              >
                <div
                  className="euiAccordion__triggerWrapper"
                >
                  <button
                    aria-controls="selectedMetricsSelector"
                    aria-expanded={true}
                    className="euiAccordion__button"
                    id="random_html_id"
                    onClick={[Function]}
                    type="button"
                  >
                    <span
                      className="euiAccordion__iconWrapper"
                    >
                      <EuiIcon
                        className="euiAccordion__icon euiAccordion__icon-isOpen"
                        size="m"
                        type="arrowRight"
                      >
                        <EuiIconEmpty
                          aria-hidden={true}
                          className="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
                          focusable="false"
                          role="img"
                          style={null}
                        >
                          <svg
                            aria-hidden={true}
                            className="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
                            focusable="false"
                            height={16}
                            role="img"
                            style={null}
                            viewBox="0 0 16 16"
                            width={16}
                            xmlns="http://www.w3.org/2000/svg"
                          />
                        </EuiIconEmpty>
                      </EuiIcon>
                    </span>
                    <span
                      className="euiIEFlexWrapFix"
                    >
                      <EuiTitle
                        size="xxxs"
                      >
                        <span
                          className="euiTitle euiTitle--xxxsmall"
                        >
                          Selected Metrics
                        </span>
                      </EuiTitle>
                    </span>
                  </button>
                </div>
                <div
                  aria-labelledby="random_html_id"
                  className="euiAccordion__childWrapper"
                  id="selectedMetricsSelector"
                  role="region"
                  tabIndex={-1}
                >
                  <EuiResizeObserver
                    onResize={[Function]}
                  >
                    <div>
                      <div
                        className=""
                      >
                        <ul
                          className="metricsList"
                        />
                      </div>
                    </div>
                  </EuiResizeObserver>
                </div>
              </div>
            </EuiAccordion>
            <EuiSpacer
              size="s"
            >
              <div
                className="euiSpacer euiSpacer--s"
              />
            </EuiSpacer>
            <EuiAccordion
              arrowDisplay="left"
              buttonContent={
                <EuiTitle
                  size="xxxs"
                >
                  <span>
                    Available Metrics
                  </span>
                </EuiTitle>
              }
              id="availableMetricsSelector"
              initialIsOpen={true}
              isLoading={false}
              isLoadingMessage={false}
              paddingSize="none"
            >
              <div
                className="euiAccordion euiAccordion-isOpen"
              >
                <div
                  className="euiAccordion__triggerWrapper"
                >
                  <button
                    aria-controls="availableMetricsSelector"
                    aria-expanded={true}
                    className="euiAccordion__button"
                    id="random_html_id"
                    onClick={[Function]}
                    type="button"
                  >
                    <span
                      className="euiAccordion__iconWrapper"
                    >
                      <EuiIcon
                        className="euiAccordion__icon euiAccordion__icon-isOpen"
                        size="m"
                        type="arrowRight"
                      >
                        <EuiIconEmpty
                          aria-hidden={true}
                          className="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
                          focusable="false"
                          role="img"
                          style={null}
                        >
                          <svg
                            aria-hidden={true}
                            className="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
                            focusable="false"
                            height={16}
                            role="img"
                            style={null}
                            viewBox="0 0 16 16"
                            width={16}
                            xmlns="http://www.w3.org/2000/svg"
                          />
                        </EuiIconEmpty>
                      </EuiIcon>
                    </span>
                    <span
                      className="euiIEFlexWrapFix"
                    >
                      <EuiTitle
                        size="xxxs"
                      >
                        <span
                          className="euiTitle euiTitle--xxxsmall"
                        >
                          Available Metrics
                        </span>
                      </EuiTitle>
                    </span>
                  </button>
                </div>
                <div
                  aria-labelledby="random_html_id"
                  className="euiAccordion__childWrapper"
                  id="availableMetricsSelector"
                  role="region"
                  tabIndex={-1}
                >
                  <EuiResizeObserver
                    onResize={[Function]}
                  >
                    <div>
                      <div
                        className=""
                      >
                        <ul
                          className="metricsList"
                        />
                      </div>
                    </div>
                  </EuiResizeObserver>
                </div>
              </div>
            </EuiAccordion>
          </section>
        </PseudoLocaleWrapper>
      </IntlProvider>
    </I18nProvider>
  </Sidebar>
</Provider>
`;
